scripts/vehicle_node.py

```diff
@@ -0,0 +1,282 @@
+        self.lock.acquire()
+        try:
+            self.twist = data
+        finally:
+            self.lock.release()
+
+
+    def update_position(self, event):
+
+        """
+        Computes the pose of the vehicle within a given footprint height and width 
+        """
+
+        #Create a copy of the most recent stored twist data to perform calculations with
+        self.lock.acquire()
+        try:
+            velocity_data = copy.deepcopy(self.twist)
+        finally:
+            self.lock.release()
+
+        #time elapsed since last update position call
+        if hasattr(event, 'last_real'):
+            if event.last_real is None:
+                time = rospy.Duration(0.05)
+            else:
+                time = event.current_real - event.last_real
+        
+        time = time.to_sec()
+
+        #Calculate angle turned in the given time using omega = theta/time
+        #TODO Currently it yaws, then moves forward. Revist calculation to do both simultaneously
+        angle = velocity_data.angular.z*time
+
+        #Calculate distance travelled in the given time using linear velocity = arc distance/time
+        distance = velocity_data.linear.x*time
+
+        #Calculate yaw of the robot
+        self.vehicle_yaw += angle
+
+        #Calculate vehicle x, y, z position coordinates
+        # TODO recalculate the position based on traveling in a circular arc.
+        self.pose.position.x += (distance)*cos(self.vehicle_yaw)
+        self.pose.position.y += (distance)*sin(self.vehicle_yaw)
+
+        ##########Calculate z position using linear interpolation and create cloud array########
+        
+        #Create range to be used in interpolation function
+        x = np.linspace(0, self.length, self.resolution)
+        y = np.linspace(0, self.width, self.resolution)
+
+        #Create cloud array to be converted to point cloud
+        cloud = []
+        for i in range(len(self.gaussian_array)):
+            for j in range(len(self.gaussian_array[i])):
+                innerlist = []
+                innerlist.append(x[i])
+                innerlist.append(y[j])
+                innerlist.append(self.gaussian_array[i][j])
+                cloud.append(innerlist)
+
+        #Create interpolation function based on the ranges and gaussian data
+        f = RectBivariateSpline(x,y, self.gaussian_array)
+
+        #Create the footprint for the vehicle
+        x1 = np.linspace((self.pose.position.x-self.vehicle_length), (self.pose.position.x+self.vehicle_length),5)
+        y1 = np.linspace((self.pose.position.y-self.vehicle_width), (self.pose.position.y+self.vehicle_width),5)
+
+        #Interpolate z values for points in the footprint of the vehicle
+        z = f(x1, y1)
+        
+        #Calculate the average of the points in the footprint and assign it to the z position of the vehicle.
+        flat_list = []
+
+        for sublist in z:
+            for item in sublist:
+                flat_list.append(item)
```

Comment:
Nit, z is more than likely a numpy array so you can just say numpy.sum(z) and this would give you the sum.